NIFI-3709: Export NiFi flow dataset lineage to Apache Atlas

ijokarumawak opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

ijokarumawak commented:
Dear reviewers. This PR has more than 20K lines of code, it may be difficult to review it quickly. However, I wrote as much documentation as possible at additionalDetail.html of the NiFiAtlasFlowLineage reporting task, and hopefully it helps to make the review cycle quicker. Please read the doc first. Thanks for your time and effort in advance!

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFiAtlasClient.java:
```diff
@@ -0,0 +1,537 @@
+        } catch (UniformInterfaceException e) {
+            if (e.getResponse().getStatus() == 204) {
+                // 204 is a successful response.
+                // NOTE: However after executing this, Atlas should be restarted to work properly.
```
Is this a bug in Atlas? If so, has a JIRA been filed for the Atlas team? Should we be logging a warning here?

ijokarumawak replied:
I reported several findings to Atlas project while I work on this integration. However, I think I haven't share this particular issue yet. Thanks, I will do. This method is only used by ITTest class, though.

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFiAtlasClient.java:
```diff
@@ -0,0 +1,537 @@
+            shouldUpdateNiFiFlow = true;
+        }
+
+        logger.debug("### NiFi Flow Audit Logs START");
```
This whole block should probably be wrapped in a check if logger.isDebugEnabled()

ijokarumawak replied:
Indeed. Thanks for the advice!

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFiAtlasClient.java:
```diff
@@ -0,0 +1,537 @@
+                final Map<String, Object> uniqueAttrs = Collections.singletonMap(ATTR_QUALIFIED_NAME, entity.getAttribute(ATTR_QUALIFIED_NAME));
+                return new Tuple<>(new AtlasObjectId(id.getGuid(), id.getTypeName(), uniqueAttrs), entity);
+            } catch (AtlasServiceException e) {
+                return null;
```
If an Exception is being thrown we should probably be at least logging it, rather than just swallowing it, no?

ijokarumawak replied:
Absolutely. Thanks!

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/NiFiProvenanceEventAnalyzerFactory.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    public static NiFiProvenanceEventAnalyzer getAnalyzer(String typeName, String transitUri, ProvenanceEventType eventType) {
+
+        if (!loaded) {
```
This is not thread-safe. This is an instance of Double-checked locking (https://en.wikipedia.org/wiki/Double-checked_locking)

ijokarumawak replied:
Thank you! I was referring an ancient blog post to implement this logic.. I will update it to use static lazy initializer.

ijokarumawak replied:
Specifically, "Initialization-on-demand holder idiom".
https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/StandardAnalysisContext.java:
```diff
@@ -0,0 +1,136 @@
+    private ComputeLineageResult getLineageResult(long eventId, ComputeLineageSubmission submission) {
+        final ComputeLineageResult result = submission.getResult();
+        try {
+            if (result.awaitCompletion(10, TimeUnit.SECONDS)) {
```
I'm not sure yet the context in which this is used. However, this is a bit of a red flag. In order to compute a FlowFile's lineage, depending on the size of the Provenance Repository and the implementation, it may take many minutes (consider a Provenance Repository that is a couple of terabytes in size and the event is one of the oldest -it will have to search all lucene indices for this). In most cases it is fairly quick, but we should not assume that it will be. What exactly is the implication here if this always times out?

ijokarumawak replied:
I agree on that this can be a costly operation. The reason to query provenance is to compute a lineage from a DROP provenance event. This is used by 'Complete Path' strategy. I wrote documentation on performance impact. If this does not work for a use-case, then user can choose another strategy, 'Simple Path'. Simple Path does not query provenance events this way. It analyzes each individual event, so should be more lightweight.

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/StandardAnalysisContext.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    // NOTE: This user is required to avoid NullPointerException at PersistentProvenanceRepository.submitLineageComputation
+    private static final QueryNiFiUser NIFI_USER = new QueryNiFiUser();
```
Using this approach means that any NiFi that is run securely will not be able to make use of this Atlas integration. I think the more correct solution is to update the Provenance Repository so that we can run a query without supplying a user. In the mean time, we could perhaps get away with allowing the user to configure the username to run the queries as? Then they can simply create an "Atlas Integration" user or choose their own user, etc.

ijokarumawak replied:
Actually this user can be anything. Since NiFI authentication and authorization is done at the web request layer, when the provenance API is called directly, no authentication nor authorization take place. I confirmed that provenance query work in secured NiFi environment. Is that approach fine?

markap14 replied:
@ijokarumawak The authentication is done at the web request layer and authorization for making a provenance query is done there, but the event-level authorization is not. That is done in the repository itself. So in a secure environment, the lineage that is returned should consist of "UNKNOWN" event types. If that is not the behavior you are seeing, then I believe we have a bug somewhere.

ijokarumawak replied:
I see. What I am seeing is different from your description. I will try to find where the event level auth is implemented.

markap14 replied:
@ijokarumawak I think I understand now what is going on. In a secured environment, you would indeed get back all UNKNOWN event types, because that user would not be authorized to see the events. But even the UNKNOWN event would include the ID of the event. It looks like you're not actually using the other info from the lineage but rather just using the Event ID to lookup the actual Event (doing so without a user). So all of this seems to me now :) Sorry about all of the confusion - and thanks for the clarifications!

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/lineage/SimpleFlowPathLineage.java:
```diff
@@ -0,0 +1,205 @@
+            }
+
+            // Set groupId from incoming connection if available.
+            final List<ConnectionStatus> incomingRelationShips = nifiFlow.getIncomingRelationShips(portProcessId);
```
I think this is meant to get incoming connections, not relationships, right?

ijokarumawak replied:
Please excuse my ignorance, but is there any difference in 'connection' and 'relationship' in NiFi terminology? Probably I'm missing something..

markap14 replied:
@ijokarumawak yes, there is a distinction. A 'relationship' is defined by a processor in order to indicate which route a FlowFile should take. A Connection is defined by a user when creating the dataflow. The connection consists of one or more Relationships, and multiple connections can consist of the same relationship. So when a processor transfers FlowFile A to relationship 'success' for example, it may be placed into a single Connection. Or cloned and added to 5 different Connections. Or it may be auto-terminated and not go into any Connections. Does that make sense?

ijokarumawak replied:
Ah, that does make sense. Then this should be 'connections'. I will rename variables. Thank you

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/reporting/AtlasNiFiFlowLineage.java:
```diff
@@ -0,0 +1,714 @@
+@DynamicProperty(name = "hostnamePattern.<ClusterName>", value = "hostname Regex patterns", description = RegexClusterResolver.PATTERN_PROPERTY_PREFIX_DESC)
+// In order for each reporting task instance to have its own static objects such as KafkaNotification.
+@RequiresInstanceClassLoading
+public class AtlasNiFiFlowLineage extends AbstractReportingTask {
```
I wonder if a more consistent name may be ReportLineageToAtlas, following with the <Verb><Noun> naming scheme

ijokarumawak replied:
Agreed. Thanks for your suggestion!

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/reporting/AtlasNiFiFlowLineage.java:
```diff
@@ -0,0 +1,714 @@
+            .name("atlas-conf-create")
+            .displayName("Create Atlas Configuration File")
+            .description("If enabled, 'atlas-application.properties' file will be created in 'Atlas Configuration Directory'" +
+                    " automatically when this processor starts." +
```
I think this was meant to read 'when this Reporting Task starts', rather than 'when this processor starts'

ijokarumawak replied:
Thank you for finding it!

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/reporting/AtlasNiFiFlowLineage.java:
```diff
@@ -0,0 +1,714 @@
+            .displayName("Kafka Bootstrap Servers")
+            .description("Kafka Bootstrap Servers to send Atlas hook notification messages based on NiFi provenance events." +
+                    " E.g. 'localhost:9092'" +
+                    " NOTE: Once this reporting task has started, restarting NiFi is required to changed this property" +
```
Has a JIRA been filed with Atlas to make this more manageable?

ijokarumawak replied:
I have shared this issue with Atlas project.

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/reporting/AtlasNiFiFlowLineage.java:
```diff
@@ -0,0 +1,714 @@
+        atlasClient = NiFiAtlasClient.getInstance();
+        try {
+            atlasClient.initialize(urls.toArray(new String[]{}), atlasAuthN, confDir);
+        } catch (final NullPointerException e) {
```
This seems odd to me, to catch a NullPointer - why would that be thrown? Should this be a more general Exception?

ijokarumawak replied:
Yeah, this is odd to me, too. However, AtlasClient throws NullPointerException when it can not find atlas properties file with following stack trace. Probably we should report this to Atlas project, too. Thanks.

```
2017-12-15 23:46:36,607 ERROR [StandardProcessScheduler Thread-4] o.a.n.c.s.StandardProcessScheduler Failed to invoke the On-Scheduled Lifecycle methods of [AtlasNiFiFlowLineage[id=5a7835aa-0160-1000-974d-14d95e0c38e9], java.lang.reflect.InvocationTargetException, 30 sec] due to {}; administratively yielding this ReportingTask and will attempt to schedule it again after {}
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotations(ReflectionUtils.java:137)
        at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotations(ReflectionUtils.java:125)
        at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotations(ReflectionUtils.java:70)
        at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotation(ReflectionUtils.java:47)
        at org.apache.nifi.controller.scheduling.StandardProcessScheduler$2.run(StandardProcessScheduler.java:210)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.nifi.processor.exception.ProcessException: Failed to initialize Atlas client due to java.lang.NullPointerException. Make sure 'atlas-application.properties' is in the directory specified with Atlas Configuration Directory or under root classpath if not specified.
        at org.apache.nifi.atlas.reporting.AtlasNiFiFlowLineage.initAtlasClient(AtlasNiFiFlowLineage.java:505)
        at org.apache.nifi.atlas.reporting.AtlasNiFiFlowLineage.setup(AtlasNiFiFlowLineage.java:402)
        ... 16 common frames omitted
Caused by: java.lang.NullPointerException: null
        at org.apache.atlas.AtlasBaseClient.getClient(AtlasBaseClient.java:171)
        at org.apache.atlas.AtlasBaseClient.initializeState(AtlasBaseClient.java:154)
        at org.apache.atlas.AtlasBaseClient.initializeState(AtlasBaseClient.java:149)
        at org.apache.atlas.AtlasBaseClient.<init>(AtlasBaseClient.java:96)
        at org.apache.atlas.AtlasClientV2.<init>(AtlasClientV2.java:115)
        at org.apache.nifi.atlas.security.Basic.createClient(Basic.java:62)
        at org.apache.nifi.atlas.NiFiAtlasClient.initialize(NiFiAtlasClient.java:119)
        at org.apache.nifi.atlas.reporting.AtlasNiFiFlowLineage.initAtlasClient(AtlasNiFiFlowLineage.java:503)
        ... 17 common frames omitted
```

ijokarumawak replied:
Since the reason of NullPointer is known (it's possible to get due to different causes, but..) for NiFi users sake, I converted the Exception to more understandable error message.

markap14 commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/resources/docs/org.apache.nifi.atlas.reporting.AtlasNiFiFlowLineage/additionalDetails.html:
```diff
@@ -0,0 +1,541 @@
+            </ul>
+        </ul>
+        <ul>
+            <!-- TODO: link to S2S details -->
```
This was intended to be resolved?

ijokarumawak replied:
Forgot to delete the TODO. I was willing to write about the detail, but it's too internal I think. So just remove the TODO comment. I found other TODOs remaining in this document. All resolved.

markap14 commented:
@ijokarumawak thanks for all of the work that you've put into this - it is very much a non-trivial effort! For the most part, the code looks good. I flagged a couple of minor things in the code, 1 or 2 thread-safety issues that should be easy to address.

The only 'more significant' concern that I have is the use of the dummied-up NiFiUser. As-is, this is an anonymous user and in a secured environment will not retrieve the event details that are necessary. It also means that we would be validating events against a user who doesn't even exist.

I think there are 2 ways to approach this: first, as I noted inline, we could have a property to define which user the queries should run on behalf of. So the user could add a "NiFi Atlas" user and use that. However, that's also a bit concerning because it means that whoever has access to edit the reporting task can run provenance queries on behalf of another user.

By far, my preference is to actually just update the ProvenanceRepository implementations (There are 4 now, I think) so that if a null User is passed in, we don't check permissions. This would mean that you can pass in null from Reporting Task. We could also then update the interface to have an overloaded method that does not require that a user be given.

Once that is addressed, I think it is a +1 from me from a code review perspective.

Thanks
-Mark

dmkoster commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFIAtlasHook.java:
```diff
@@ -0,0 +1,294 @@
+ */
+public class NiFIAtlasHook extends AtlasHook implements LineageContext {
+
+    public static final String NIFI_USER = "nifi";
```
Would this be best offered as a configurable value?

ijokarumawak replied:
Yes, it would. However, at this moment, this user is not significant. Please see this conversation on the same subject.
https://github.com/apache/nifi/pull/2335#discussion_r156992235

dmkoster commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFIAtlasHook.java:
```diff
@@ -0,0 +1,294 @@
+
+
+    private static <K, V> Map<K, V> createCache(final int maxSize) {
+        return new LinkedHashMap<K, V>(maxSize, 0.75f, true) {
```
The default load factor is currently 0.75. Is there a reason for fixing the value versus allowing future default optimizations?

ijokarumawak replied:
Well, it's just a private method to create cache instances within this class. So if different default optimizations are needed, then we can do that here as well without affecting others.

dmkoster commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFiAtlasClient.java:
```diff
@@ -0,0 +1,537 @@
+                final EntityMutationResponse mutationResponse = atlasClient.createEntities(atlasEntities);
+                logger.debug("mutation response={}", mutationResponse);
+            } catch (AtlasServiceException e) {
+                if (e.getStatus().getStatusCode() == 404 && e.getMessage().contains("ATLAS-404-00-00B")) {
```
Can this reference AtlasErrorCode.INSTANCE_NOT_FOUND directly instead of hardcoding "ATLAS-404-00-00B"?

ijokarumawak replied:
That is a good idea. Thanks!

dmkoster commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/lineage/SimpleFlowPathLineage.java:
```diff
@@ -0,0 +1,205 @@
+
+            // Create a queue.
+            Referenceable queueFromStaticFlowPathToRemotePortProcess = new Referenceable(TYPE_NIFI_QUEUE);
+            queueFromStaticFlowPathToRemotePortProcess.set(ATTR_NAME, "queue");
```
There's a few places that "queue" is hardcoded. Should it be added to NiFiTypes?

ijokarumawak replied:
Hmm, I think it should be ok as it is, since it's a 'name' property. It does not have to be consistent across different context, if it's 'typeName' or 'qualifiedName' or any other important attribute, then it probably should be a constant, but it's just a name. Thanks anyway.

ijokarumawak commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/AtlasUtils.java:
```diff
@@ -0,0 +1,77 @@
+
+
+    public static boolean isGuidAssigned(String guid) {
+        return guid != null && !guid.startsWith("-");
```
Yes, it's Atlas internal implementation. However, checking null is not sufficient, because entities created at client side but not yet registered in Atlas have negative GUIDs. I'd like to keep it this way if there's no strong objections. Thank you.

ijokarumawak commented on nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFIAtlasHook.java:
```diff
@@ -0,0 +1,294 @@
+ * This class is not thread-safe as it holds uncommitted notification messages within instance.
+ * {@link #addMessage(HookNotificationMessage)} and {@link #commitMessages()} should be used serially from a single thread.
+ */
+public class NiFIAtlasHook extends AtlasHook implements LineageContext {
```
Good catch, thank you! I'll fix this.

ijokarumawak commented:
@markap14 @dmkoster Thank you very much for your informative review feedback! I've addressed all comments other than confirming the Provenance Event level authorization. I will keep digging that.

ijokarumawak commented:
@markap14 I found that ReportLineageToAtlas does get UNKNOWN provenance events when it queries lineage by `ProvenanceRepository.submitLineageComputation(final long eventId, final NiFiUser user)`. It uses the event hierarchies returned by that method, but to analyze each event detail fully, it also calls `ProvenanceEventRepository.getEvent(final long id)` which does not require user. That's why the reporting task worked even in a secured NiFi cluster.

I now understand why NiFi framework has two interfaces for provenance events, i.e. ProvenanceRepository and ProvenanceEventRepository. `ProvenanceRepository` provides methods to be called in a context with a user who made the request. `ProvenanceEventRepository` is used outside of a user request, intended for Reporting Tasks, etc.

Is my understanding correct? If so, the approach of this PR looks good now? Thanks!

markap14 commented:
@ijokarumawak yes I believe your understanding to be correct. Thanks for the clarifications! In terms of code I am good with the PR. Just want to test this against a live instance and as soon as I can confirm that all works well, I can push to master.

However, I do see that this nar alone is > 60 MB. For now, I think I it would be best to add a Maven profile for activating this so that those who don't need this don't have to pull in the extra weight. Is that OK with you?

markap14 commented:
@ijokarumawak all looks good after some local testing. Thanks for your diligence and the willingness to keep going on this until we got everything resolved! I have now merged this to master!
